packages/dashboard/src/DashboardLayout.tsx

```diff
@@ -0,0 +1,254 @@
+
+const EMPTY_OBJECT = Object.freeze({});
+
+const DEFAULT_LAYOUT_CONFIG: DashboardLayoutConfig = [];
+
+const DEFAULT_CALLBACK = () => undefined;
+
+type DashboardData = {
+  closed?: ClosedPanels;
+};
+interface DashboardLayoutProps {
+  id: string;
+  layout: GoldenLayout;
+  layoutConfig?: DashboardLayoutConfig;
+  onLayoutChange?: (dehydratedLayout: DashboardLayoutConfig) => void;
+  data?: DashboardData;
+  children?: React.ReactNode | React.ReactNode[];
+}
+
+/**
+ * DashboardLayout component. Handles hydrating, dehydrating components, listening for dragging panels.
+ */
+export const DashboardLayout = ({
+  id,
+  children,
+  layout,
+  layoutConfig = DEFAULT_LAYOUT_CONFIG,
+  onLayoutChange = DEFAULT_CALLBACK,
+}: DashboardLayoutProps): JSX.Element => {
+  const dispatch = useDispatch();
+  const data =
+    useSelector(state => getDashboardData(state, id)) ?? EMPTY_OBJECT;
+
+  const [isItemDragging, setIsItemDragging] = useState(false);
+  const [lastConfig, setLastConfig] = useState<DashboardLayoutConfig>();
+  const [initialClosedPanels] = useState(data?.closed ?? []);
+
+  const hydrateMap = useMemo(() => new Map(), []);
+  const dehydrateMap = useMemo(() => new Map(), []);
+  const store = useStore();
+  const registerComponent = useCallback(
+    (
+      name: string,
+      componentType: ComponentType,
+      hydrate = hydrateDefault,
+      dehydrate = dehydrateDefault
+    ) => {
+      log.debug2('registerComponent', name, componentType, hydrate, dehydrate);
+
+      function renderComponent(
+        props: { glContainer: unknown; glEventHub: unknown },
+        ref: unknown
+      ) {
+        // Cast it to an `any` type so we can pass the ref in correctly.
+        // ComponentType doesn't seem to work right, ReactNode is also incorrect
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        const CType = componentType as any;
+
+        // Props supplied by GoldenLayout
+        // eslint-disable-next-line react/prop-types
+        const { glContainer, glEventHub } = props;
+        return (
+          <Provider store={store}>
+            <PanelErrorBoundary
+              glContainer={glContainer}
+              glEventHub={glEventHub}
+            >
+              {/* eslint-disable-next-line react/jsx-props-no-spreading */}
+              <CType {...props} ref={ref} />
+            </PanelErrorBoundary>
+          </Provider>
+        );
+      }
+
+      const wrappedComponent = React.forwardRef(renderComponent);
+      const cleanup = layout.registerComponent(name, wrappedComponent);
+      hydrateMap.set(name, hydrate);
+      dehydrateMap.set(name, dehydrate);
+      return cleanup;
+    },
+    [hydrateMap, dehydrateMap, layout, store]
+  );
+  const hydrateComponent = useCallback(
+    (name, props) => hydrateMap.get(name)?.(props, id),
+    [hydrateMap, id]
+  );
+  const dehydrateComponent = useCallback(
+    (name, props) => dehydrateMap.get(name)?.(props, id),
```

Comment:
We need to try to dehydrate the components that are not in the map. Maybe add a fallback here.. or everywhere `dehydrateComponent` is called.